File: backend/src/utils/audit.ts
import { AuditLog } from '@database/index';
import { getRequestContext } from '@utils/context';
import { Transaction } from 'sequelize';

export interface AuditEntry {
  actorId?: number | null;
  action: string;
  module: string;
  entityType?: string | null;
  entityId?: number | null;
  oldValues?: object | null;
  newValues?: object | null;
}

export async function writeAuditLog(entry: AuditEntry, transaction?: Transaction) {
  const { ip, userAgent } = getRequestContext();

  await AuditLog.create(
    {
      userId: entry.actorId ?? null,
      action: entry.action,
      module: entry.module,
      entityType: entry.entityType ?? null,
      entityId: entry.entityId ?? null,
      oldValues: entry.oldValues ?? null,
      newValues: entry.newValues ?? null,
      ipAddress: ip ?? null,
      userAgent: userAgent ?? null,
    },
    { transaction },
  );
}

File: backend/app/core/audit.py
from app.core.context import get_context
from app.models import AuditLog


def write_audit(db, actor_id, action, module, entity_type=None, entity_id=None, old=None, new=None):
    ctx = get_context()
    db.add(AuditLog(
        userId=actor_id,
        action=action,
        module=module,
        entityType=entity_type,
        entityId=entity_id,
        oldValues=old,
        newValues=new,
        ipAddress=ctx['ip'],
        userAgent=ctx['user_agent'],
    ))